### ISSUE TITLE
Harbor exausts postgresql connections

### ISSUE THREAD

**Issue by ricardojdsilva87** [goharbor/harbor#16915]
Hello,
We have an Harbor instalation with the following settings:
- Harbor v2.5.0
- External Postgresql database hosted as RDS on AWS
- External Redis database on AWS
- 3 replicas per service

We encountered an issue when putting more load into Harbor where all the connections to the database where exhausted and it killed the service.
We are using also Harbor as a proxy to dockerhub so a few of the calls need to be served by Harbor when a node comes up. When putting more images to be hosted and pulled from Harbor, basically the number of connections to the database grew until the 3000 (limit of an m5.2xlarge RDS with 32GB of RAM).

We find this odd since we just started using Harbor, the "normal" number of connections are around 250 or so, but when more load is put into Harbor, basically for new images to be pulled/pushed, the DB crashes.

We don't see that increasing vertically the underlying RDS machine is a solution and tried to use an Aurora RDS Postgres with multiple read replicas. The issue here is that there is no place to configure a read replica (does harbor support it?). After the change the main writer instance kept being used and all the connections were made there.

On the helm chart we have only this option to configure related to postgres:
![image](https://user-images.githubusercontent.com/12953702/170519814-bca386bc-93e5-428b-8bd2-5586142fe1e1.png)

Here are the graphs for the sessions and number of commits:
<img width="1383" alt="image" src="https://user-images.githubusercontent.com/12953702/170520365-89701df7-de42-4cf1-828c-c8d784279eb0.png">

Here are the waits, most of them of the type lock:tuple
<img width="1349" alt="image" src="https://user-images.githubusercontent.com/12953702/170520611-52d2ea59-d973-485e-a52e-99399739b5db.png">

And the TOP SQL queries:
<img width="1424" alt="image" src="https://user-images.githubusercontent.com/12953702/170520815-b8e0cb40-965e-43be-a541-f76aef799c08.png">

In this case it seems that an update of the quota is also locking the database:
<img width="751" alt="image" src="https://user-images.githubusercontent.com/12953702/170521129-e0d798e1-4e85-43e4-b328-4d27e4c46d12.png">

The select showed in second place could be one of the candidates to be used by a read replica:
<img width="702" alt="image" src="https://user-images.githubusercontent.com/12953702/170521243-88d08fb5-ccde-4d0d-a040-4dc89fc7f152.png">

Let me know if you need more information.
Thanks for the help

**Comment by chlins** [comment]
Currently harbor does not support configure postgres read replica, the sql shown above is trying to update quota when client push image, so did you meet this issue when you have concurrent push to one or different project and how many clients about?

**Comment by chlins** [comment]
Do you have any outside periodic job or configure any schedule replication or scan job in harbor?

**Comment by chlins** [comment]
You also could list audit logs to figure out who and when make these operations.

**Comment by ricardojdsilva87** [comment]
Hello @chlins ,
The only job present is the garbage collection, triggered every day at midnight, but we don't see any impact from that.
About the user all the calls are made by the configured RDS user on the Harbor deployment.

**Comment by ricardojdsilva87** [comment]
Hello @chlins ,
On the 27th around 3AM we have some fewer calls, mostly calling for assets for the docker registry mirror configured on harbor:
![image](https://user-images.githubusercontent.com/12953702/170954690-c5979951-5c6a-4cdb-9199-a395526c605e.png)

Basically some HEAD and GET requests to fetch the assets.
```
[27/May/2022:02:51:27 +0000] "HEAD /v2/library/busybox/manifests/stable?ns=docker.io HTTP/1.1" 200 0 "-" "containerd/v1.5.8" 1678 0.209
[27/May/2022:01:46:25 +0000] "GET /service/token?scope=repository%3Alibrary%2Fbusybox%3Apull&scope=repository%3Aproxy%2Flibrary%2Fbusybox%3Apull&service=harbor-registry HTTP/1.1" 200 959
```

And Harbor serving the blob from S3:
```
[27/May/2022:01:31:25 +0000] "GET /v2/selenium/node-chrome/blobs/sha256:234b765785847b0b42949b7851aa6acde5bd7f1963ee9b8f888f1cafa90f6797?ns=docker.io HTTP/1.1" 307 0 "-" "containerd/v1.5.8" 1639 0.230
```

Not sure what triggers all the UPDATE quota requests, but I'm thinking that maybe the behaviour that we've seen on the proxy cache where we had several image tags that supposed weren't requested might have triggered that temporarly

We are seeing this at the moment, some locks:
![image](https://user-images.githubusercontent.com/12953702/170955709-81277b57-365d-4ae6-96bc-be2724288405.png)

Checking the logs there aren't too much requests happening reaching the ingress controller and thus Harbor.
![image](https://user-images.githubusercontent.com/12953702/170955808-ba7aec69-450c-4485-a904-774648f1c894.png)

**Comment by chlins** [comment]
Harbor also records pull/push actions from outside, could you check that logs to catch some mass pull/push?<img width="1849" alt="image" src="https://user-images.githubusercontent.com/31262637/171084000-0ba0d913-c298-4020-bc7b-918e7ec96844.png">

**Comment by ricardojdsilva87** [comment]
Hello @chlins ,
Yes we have some pulls made around the same time, but it is considered normal since we have the EC2 nodes for the containers we use (ephemeral) just being spawned and killed, there can be 10/20 to spin up at a certain time.
Also remembered 1 important aspect, we have PrismaCloud as a CVE checker integrated with Harbor as the scanner and we had the agents scanning the images every 2 hours and changed that to 24 hours timeframe.
From there we saw the following with the locks:
![image](https://user-images.githubusercontent.com/12953702/171196843-4dc7d21e-11c5-4894-95e9-e2ac5028062e.png)
![image](https://user-images.githubusercontent.com/12953702/171196881-99190aa0-5e8e-48d4-8c02-c9a3d4234cc6.png)
The lock happens every 24 hours now.
It seems that even though the PrismaCloud agents just pull the images from Harbor, it seems that there is also a quota update? Shouldn't that be only for push requests when an image is created?
Thanks for the help

**Comment by chlins** [comment]
I think pull image will not update quota, could you confirm the lock type to determinate whether the lock is posed by `update quota` or other db query?

**Comment by chlins** [comment]
@ricardojdsilva87 Hi, we have tried to reproduce this issue by enable the schedule scan, but from our result no connection spike and db update quota lock occurred, so could you provide more information, did the update quota lock happened every scan interval? Do you have any job to list artifacts or catalog from proxy cache project and then pull them?(or maybe PrismaCloud will have this behavior?)

**Comment by ricardojdsilva87** [comment]
Hello @chlins ,
I have checked and this is want the last week looks like:
![image](https://user-images.githubusercontent.com/12953702/172325227-1f356e42-ced1-47c5-8514-577639fed001.png)
![image](https://user-images.githubusercontent.com/12953702/172325291-48c0c694-5baf-444c-950f-c5baf1fc5c1f.png)
![image](https://user-images.githubusercontent.com/12953702/172325363-ce4740c1-45d9-4c0a-aa41-902762ae8403.png)
It doesn't look as bad as it was some weeks ago. It seems that scheduling the prismacloud scan to once a day also helped.
I can see some other queries on the TOP SQL that might be related to scanning, mostly because I think that the result of the CVE's need to be saved for each image on the database correct?
As I can see the scan is triggered for all images every day around 9:00 UTC.
The queries seen are:
```
UPDATE quota_usage SET used = $1, update_time = $2, version = $3 WHERE id = $4 AND version = $5
INSERT INTO "audit_log" ("project_id", "operation", "resource_type", "resource", "username", "op_time") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"
UPDATE artifact SET pull_time = $1 WHERE id = $2 AND (pull_time IS NULL OR pull_time < $3)
```
The Update quota one is in fact the one that impacts the most the database, but I'm not sure if it's triggered somehow by the other 2, for the audit_log and also the SET pull_time that is also updated.

Thanks

**Comment by chlins** [comment]
@ricardojdsilva87 When harbor trigger a scan job, it will list all artifacts and then scan them, and the scanner needs to pull artifacts from harbor, so the sql of `insert audit_log` and `update pull_time` match to expect, but the `update quota` should only happened when push artifact to harbor. So could you export the recently harbor audit logs(if you have any security concern, could hide some sensitive info)? and then we can help to analyze these logs to find out the root cause.

**Comment by ricardojdsilva87** [comment]
Hello @chlins ,
Sorry for the delay. Here is the data for today:
<img width="1643" alt="image" src="https://user-images.githubusercontent.com/12953702/172819812-c0658727-f8bf-485d-bde1-3af7d520636c.png">
Looking to the audit logs I can see that practically around that time we have several create artifacts for the docker proxy we use (to prevent the API limit from docker hub)
<img width="1465" alt="image" src="https://user-images.githubusercontent.com/12953702/172820229-23d7954e-33bf-442e-9974-0565b9a0d2cf.png">
We've seen a strange behaviour and I've already deleted 1 or 2 repositories from the dockerproxy one.
An example we have is that for sonarqube we are just using 1 version and the existent artifacts ranged from the first versions until alpha, beta, etc, so around 30 artifacts were downloaded at the same time. Checking our test Harbor instance we just have the versions we pull
Test harbor:
<img width="769" alt="image" src="https://user-images.githubusercontent.com/12953702/172820683-4401709c-2570-4adc-bedb-9bb1806555fe.png">
This behaviour started to happen after the issues we started having, altought it doesn't make sense. Some registries just have alpha, beta, RC and arm releases that we do not even use...
I can try to check some of these repositories and delete them if we see that there are too many images that simply do not make any sense to exist. This way if a new request is done, harbor would just get the image from dockerhub and recreate the repository.
Thanks

**Comment by chlins** [comment]
Hi, the strange behavior as you described is when you only pull one tag like `v1` but get many unrelated tags like `v2, v3...`, so where you pull these expect artifacts, did you pull them by manually or any other replication jobs?

**Comment by ricardojdsilva87** [comment]
Yes that's it. it seems that some of the repos, not all show the same strange behaviour.
This repository uses the docker proxy feature of Harbor. They are pulled into harbor registry cache if they are requested in the kubernetes cluster.
Imagine that you need busybox:1.2.3 you mention it on your deployment and we are setting the registry mirror option on containerd to point to the harbor URL proxy cache.
This will check if the image exists locally on Harbor if so it serves it directly if not it will pull it from dockerhub and store it locally. This prevents us of reaching the dockerhub API rate limit by serving the images from Harbor directly.
We would expect that just a few image tags appear on this registry, for example for busybox this appears to be a normal behaviour:
<img width="544" alt="image" src="https://user-images.githubusercontent.com/12953702/172837124-eeb6963e-952e-45ae-88f8-48fd93c60e98.png">
For the following image for example we don't expect to have the rc versions for example:
<img width="499" alt="image" src="https://user-images.githubusercontent.com/12953702/172837753-d155366b-b759-4a4b-8fb7-be80c673da66.png">

Thanks

**Comment by chlins** [comment]
From harbor side it will only cache the tag which be pulled from outside, so these unexpected tags maybe pulled by any other clients. I think you can go through the nginx requests logs or harbor audit log to find whether it has the records that the unexpected tags requests from outside. But for this issue as you described, the root cause is the concurrency pull requests to harbor proxy cache project and for harbor it needs to pull them from dockerhub and then push to harbor, but push image to harbor one project needs to update quota which has lock so cause the db lock, waits and connection full.

**Comment by ricardojdsilva87** [comment]
Hello @chlins ,
Even after deleting a few projects from the proxy cache it seems that they appear again. This one for example was created on June 12th:
<img width="1203" alt="image" src="https://user-images.githubusercontent.com/12953702/174567520-1f84bd28-7bc3-4005-9e87-52cc840bef2d.png">
We just use 1 version of the container so does it makes sense pulling all available images from dockerhub? is there any possible configuration we are missing?
Thanks

**Comment by ricardojdsilva87** [comment]
Hello @chlins ,
Just an update, I think that I've found what is causing the artifacts to be downloaded on the Harbor docker proxy.
It seems that Prismacloud retains the artifacts scan results for x time so next time it triggers a scan it will try to pull the image from the information already retained on it. Since Harbor receives a pull request to a docker image that is not already present on Harbor (manually deleted) the way that the proxy works it pulls again the image from dockerhub because Prisma requested it. 
The fact that the container was deleted from Harbor does not remove the scan information from Prisma.
We've adjusted that already and have adjusted the cleanup window to be smaller than the scan window on Prisma. This way the images disappear from Prisma and only the existent ones are scanned again.

I think that ticket can be closed since we can manage this on our side.
Thanks again for your help!
